### PR TITLE
Use `reg` instead of `wmic` (fixes #10)

### LINKED ISSUE #10: 'wmic' command throwing access denied in azure app service
I am trying to get the archive link from my node js application.This is working fine in my local machine but i am getting error in azure private app service.

octokit.repos.getArchiveLink({owner:owner,repo:sourceRepoName,archive_format:'zipball',ref:branch}) on executing this command in azure private app service I am getting the following error

{ Error: Command failed: wmic os get Caption
Access is denied.
at makeError (D:\home\site\wwwroot\node_modules\execa\index.js:172:9)
at Function.module.exports.sync (D:\home\site\wwwroot\node_modules\execa\index.js:341:15)
at module.exports.release (D:\home\site\wwwroot\node_modules\windows-release\index.js:34:24)
at module.exports (D:\home\site\wwwroot\node_modules\os-name\index.js:39:18)
at getUserAgentNode (D:\home\site\wwwroot\node_modules\universal-user-agent\index.js:6:51)
at restEndpoint (D:\home\site\wwwroot\node_modules@octokit\rest\lib\endpoint\index.js:27:48)
at restRequest (D:\home\site\wwwroot\node_modules@octokit\rest\lib\request\index.js:7:26)
code: 1,
stdout: '',
stderr: 'Access is denied.\r\n',
failed: true,
signal: null,
cmd: 'wmic os get Caption',
timedOut: false }

So on diving deep into the error I noticed that the wmic command in this package is throwing this error.
I contacted with azure team and they said that windows commands execution is denied in azure app service.
Can you provide any alternative for this as this was working fine when I was using your earlier module 'win-release'.

### PR DESCRIPTION
**Pull request by MarkTiedemann**
Decided to go for a breaking change by returning a Promise since all recent Windows versions are server/client ambigious (which requires spawning to distinguish).

### CODE REVIEW

**Comment by sindresorhus** [comment]
Then we need both an async and sync API as my existing use of this module has to be sync.

**Comment by sindresorhus** [comment]
Can you use Travis instead of AppVeyor?

**Comment by MarkTiedemann** [comment]
> Can you use Travis instead of AppVeyor?

Didn't know Travis has Windows support now. That's great news!

> Then we need both an async and sync API as my existing use of this module has to be sync.

I already implemented that, but I'm not sure that an async / sync design is the best way to go here.

If you know you are running on client Windows, then you can get the name synchronously. If you know you are running on Windows Server, then you can also get the name synchronously. The only part that requires async is determining whether you're running on client or server.

So I'm thinking the 95% use-case here should probably stay sync and just lookup the name from the Map. And we could add an async only function to determine whether you're running on client or server for those people that want to distinguish.

So my design would be:

```js
const windowsRelease = require('windows-release')

windowsRelease()
//=> 10

if (await windowsRelease.isServer()) {
  windowsRelease.server()
  //=> Server 2016
}
```

What do you think about it?

EDIT: Then again, this puts a higher mental burden on everyone using this module. For example, I myself just forgot to update the tests to account for Travis running on Windows Server...

So what's more important here? Performance or ease of use?

EDIT2: The simplified API could look something like this:

```js
const windowsRelease = require('windows-release')

// On client
await windowsRelease()
//=> 10

// On server
await windowsRelease()
//=> Server 2016
```

But even with this design, I think it's helpful to have a function to distinguish between client and server, for example, for tests that need to distinguish between being run locally on a developer machine or on a server machine in CI.

```js
// As a new function of this module:
const windowsRelease = require('windows-release')
await windowsRelease.isServer()
//=> true/false

// Or as a new module:
const isWindowsServer = require('is-windows-server')
await isWindowsServer()
//=> true/false
```

EDIT3: How about this? I think this covers all possible use-cases.

```js
const windowsRelease = require('windows-release')

// On client
// (1) release detection
await windowsRelease() //=> 10
windowsRelease.sync() //=> 10
// (2) client/server detection
await windowsRelease.isServer() //=> false
windowsRelease.isServerSync() //=> false

// On server
// (1) release detection
await windowsRelease() //=> Server 2016
windowsRelease.sync() //=> Server 2016
// (2) client/server detection
await windowsRelease.isServer() //=> true
windowsRelease.isServerSync() //=> true

// (3) version->name lookup
windowsRelease.client() //=> 10
windowsRelease.server() //=> Server 2016
```

**Comment by MarkTiedemann** [comment]
On another note, Windows Server 2019 is out and this module is not supporting it yet... Perhaps instead of using a simple mapping, we should read the friendly name dynamically (from an API, the registry, or wherever) to make this module more future-proof (or at least do so if an unknown, future version is supplied).

**Comment by MarkTiedemann** [comment]
I think it might be time to create a binary that produces some well formatted Windows release information...

For example:

```json
{
  "version": "10.0.16299",
  "release": "1709",
  "server": false,
  "desktop": true,
  "container": false,
  "name": "10",
  "edition": "Pro"
}
```

```json
{
  "version": "10.0.17763",
  "release": "1809", 
  "server": true,
  "desktop": true,
  "container": false,
  "name": "2019",
  "edition": "Datacenter"
}
```
```json
{
  "version": "10.0.17763",
  "release": "1809", 
  "server": true,
  "desktop": false,
  "container": true,
  "name": "2019",
  "edition": "Nano"
}
```
(Note that Windows Server 2019 can run with or without a GUI so both the server and the desktop property are needed to distinguish between them. There's also "Nano" and "IoT Core" editions that run in containers...)

EDIT: I think I'm getting too deep into the rabbit hole...

File: package.json [hunk 1]
```diff
@@ -12,11 +12,15 @@
 	"engines": {
 		"node": ">=6"
 	},
+	"os": [
+		"win32"
+	],
```
Comment (sindresorhus) [inline package.json]: I prefer a runtime check, as this will prevent a package to support multiple systems where this package is support for one of those.

**Comment by sindresorhus** [comment]
The main point of this module is to just get a human-friendly name of the Windows release, so we should not dig ourselves too far into the rabbit hole. If you want, I encourage you to create a more comprehensive module for gathering more info, which this module could depend on. But for my needs, I only need the name. But yes, we could expose a `isServer()` method too for convenience.

**Comment by sindresorhus** [comment]
> (Note that Windows Server 2019 can run with or without a GUI so both the server and the desktop property are needed to distinguish between them

That is still a server to me. You can also add a server to Windows XP. It's still the desktop version.

**Comment by sindresorhus** [comment]
These sounds good to me:

- `windowsRelease()`
- `windowsRelease.sync()`
- `windowsRelease.isServer()`
- `windowsRelease.isServerSync()`

**Comment by MarkTiedemann** [comment]
> The main point of this module is to just get a human-friendly name of the Windows release, so we should not dig ourselves too far into the rabbit hole.

I think there are multiple use-cases (and issues) here.

1. Providing a mapping between a Windows version and a human-friendly name (e.g. `5.1` -> `XP`). This could be maintained easily in a Map and would have to be updated whenever a new Windows version releases. This doesn't work for servers, however: `10.0` could be both `Server 2016` or `Server 2019`.

2. Determining a human-friendly name via Windows APIs / registry / other means. For example, I found this well supported registry key for getting a human-friendly name:

```batch
reg query "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion" /v ProductName

:: On my local Windows machine
Windows 10 Pro N

:: In AppVeyor
Windows Server 2012 R2 Datacenter

:: In a Windows Nano Server container
Windows Server 2016 Datacenter

:: In a Windows Server Core container
Windows Server 2016 Datacenter
```

I think the 1. use case of providing a simple mapping doesn't have great value compared to the 2. one.

Would you be fine with using the `ProductName` registry key?

PS: The performance for the `reg query` seems to be very good as well: `Process Time: 0:00:00.015`.

**Comment by MarkTiedemann** [comment]
> But yes, we could expose a `isServer()` method too for convenience.

I just noticed `is-windows-server.exe` doesn't work in Windows containers...

I found another registry key that could help with this, however:

```batch
reg query "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion" /v InstallationType

:: On my local Windows machine
Client

:: In AppVeyor
Server

:: In a Windows Nano Server container
Nano Server

:: In a Windows Server Core container
Server Core
```

So perhaps this could be implemented by checking whether it returns "Client", else it is a server (or container).

But then again, if we use the `ProductName` for the friendly name, we don't even need to distinguish between client and server.

PS: Instead of executing `reg`, we could execute `path.join(process.env.SystemRoot, 'System32', 'reg.exe')` (`C:\Windows\System32\reg.exe`).

**Comment by MarkTiedemann** [comment]
CI is failing because, on Travis (contrary to AppVeyor or Windows containers), the ProductName is `Windows Server $Edition` instead of `Windows Server $Year $Edition`... I'll write a work-around tomorrow.

**Comment by sindresorhus** [comment]
The use-case of this module is for https://github.com/sindresorhus/os-name, which does accept `os.release()` as argument too, so I'll still need to be able to translate kernel release number => product name. And it should be consistent with and without using that argument, so not sure how useful the registry key is...

**Comment by sindresorhus** [comment]
I could also stop supporting that argument in `os-name`. Not sure how much it's used or what it's used for though. I've personally never needed it, but it's hard to predict.

**Comment by MarkTiedemann** [comment]
> I could also stop supporting that argument in `os-name`. Not sure how much it's used or what it's used for though. I've personally never needed it, but it's hard to predict.

I'm gonna keep the argument. The breaking change would be to make this module async by default and to offer a `.sync([release])` function for anyone that needs it.

I won't have access to my Windows machine until Friday again. So it might take a while for me to make progress on this PR, even though I've planned the design already.

PS: Don't mind the current code; I'll have to throw away most of it.

**Comment by sindresorhus** [comment]
Sounds good. No rush :)

**Comment by MarkTiedemann** [comment]
@sindresorhus Are you okay with converting this project to TypeScript?

I managed to run the `reg` queries on 5 different Windows servers and containers using Travis and Appveyor:

- https://travis-ci.org/MarkTiedemann/win-ci-playground/builds/508159495
- https://ci.appveyor.com/project/MarkTiedemann/win-ci-playground/builds/23174021/job/m1mv3tm2m8k43wg6
- https://ci.appveyor.com/project/MarkTiedemann/win-ci-playground/builds/23174021/job/2idrkqj005in395n

Would you be okay with adding AppVeyor for this project as well?

**Comment by sindresorhus** [comment]
> Are you okay with converting this project to TypeScript?

Converting to TS requires a lot of useless config and boilerplate. I'm not sure it's worth it for such a tiny module.

> Would you be okay with adding AppVeyor for this project as well?

I would prefer not to. I have a very strong dislike for AppVeyor.

**Comment by MarkTiedemann** [comment]
> Converting to TS requires a lot of useless config and boilerplate. I'm not sure it's worth it for such a tiny module.

What other config do I need except the `tsconfig.json`? Do you have a certain preferred structure for TS repos?

> I would prefer not to. I have a very strong dislike for AppVeyor.

How come? AppVeyor seems to be much faster than Travis for Windows builds.

I also tried out Azure Pipelines (which is free for Open Source, of course). They offer the largest variety of Windows servers and containers. Would you like to try out Azure Pipelines?

I think the greater the variety of Windows servers and containers we can test this module on, the better.

PS: If `reg` doesn't work, I'll put a fallback to `wmic`.

**Comment by sindresorhus** [comment]
> What other config do I need except the tsconfig.json? Do you have a certain preferred structure for TS repos?

Yes, TS support will require extra config for AVA, XO, tsconfig. It also requires a compile-step which slows everything down. I'm just not ready to go full in on TS. It has too many quirks and limitations. I'm personally waiting it out.

> How come? AppVeyor seems to be much faster than Travis for Windows builds.

AppVeyor has always been flaky. I've wasted countless on hours on it through the years. Also awful web UI and verbose config format. I honestly rather have super slow builds than have to deal with AppVeyor.

> I also tried out Azure Pipelines (which is free for Open Source, of course). They offer the largest variety of Windows servers and containers. Would you like to try out Azure Pipelines?

Not really. It requires a very verbose config. And I've also heard it's not that great: https://toxicbakery.github.io/vsts-devops/microsoft-devops-ci/

> I think the greater the variety of Windows servers and containers we can test this module on, the better.

I don't have time to deal with all the config boilerplate and CI weirdness, but if you want to take responsibility of dealing with it (?), then I don't mind.

**Comment by MarkTiedemann** [comment]
> [TS] also requires a compile-step which slows everything down. (...) I'm personally waiting it out.

Personally, I'd say running `tsc --watch` has sped up my coding rather than slowing it down, especially with strict mode catching most bugs right when I type them. But I understand if you want to keep this module pure JS. I guess I'll just finish my PR in TS, and then run `xo --fix` over the generated `*.js`.

Should I keep the `index.d.ts` file?

> AppVeyor has always been flaky. (...) I've also heard [Azure Pipelines]'s not that great.

Guess I'm lucky I never had any trouble with either one of them.

> [I]f you want to take responsibility of dealing with it (?), then I don't mind.

Yeah, I'll take responsibility for that. In fact, I want to update the CI setup for my Windows binaries (support more architectures, support Windows containers, etc.). Once I'm done, I could apply the learnings from that to this module, if necessary.

I'm working on making a list of all known Windows versions. I already created one for Windows 10 (see [`versions.csv`](https://github.com/MarkTiedemann/windows10-versions/blob/master/versions.csv)). If I finish that, then we won't even have to spawn `reg` or `wmic` anymore (except perhaps as a fallback for future versions).

... This PR is really turning into a side job. ^_^)

**Comment by MarkTiedemann** [comment]
Another idea: Prebuilding `x64` and `ia32` native addons with N-API. The addons could be used for distinguishing client and server as well as for reading registry values.

And again, if something fails with the native addons, we can fall back to spawning.

<details>
<summary>example-addon</summary>

```cpp
#include <napi.h>
#include <windows.h>

Napi::Object Init(Napi::Env env, Napi::Object exports)
{
  OSVERSIONINFOEX osvi;
  SecureZeroMemory(&osvi, sizeof(OSVERSIONINFOEX));
  osvi.dwOSVersionInfoSize = sizeof(OSVERSIONINFOEX);
  GetVersionEx((LPOSVERSIONINFO)&osvi);
  exports.Set(Napi::String::New(env, "isServer"),
              Napi::Boolean::New(env, osvi.wProductType != VER_NT_WORKSTATION));
  return exports;
}

NODE_API_MODULE(addon, Init)
```

```js
console.log(require('./addon_x64.node'));
//=> { isServer: false }
```

</details>

**Comment by MarkTiedemann** [comment]
I've created a prebuilt N-API module for distinguishing between Windows server and client: [MarkTiedemann/is-windows-server](https://github.com/MarkTiedemann/is-windows-server))

It's prebuilt for N-API `v1-4` (both `ia32` and `x64`), tested against Node `v4-11`, and includes `reg` and `wmic` fallbacks.

@sindresorhus Are you fine with using prebuilt N-API modules in this (and other) Windows module(s)?

CI isn't in-place yet, but I'll work on it next.

**Comment by sindresorhus** [comment]
> Should I keep the index.d.ts file?

Yup

> Are you fine with using prebuilt N-API modules in this (and other) Windows module(s)?

As long as there's a fallback, sure. I just know that even with N-API, native add-ons have a tendency to cause a lot of problems and support requests.

> ... This PR is really turning into a side job. ^_^)

LOL. I know that feeling. Deep down the rabbit hole.

**Pull request closed without merging**